src/java/com/puppetlabs/jruby_utils/pool/JRubyPool.java

```diff
@@ -134,6 +140,9 @@ public E borrowItem() throws InterruptedException {
             do {
                 if (isPoolLockHeldByAnotherThread(currentThread)) {
                     poolNotLocked.await();
+                } else if (this.poisonPill != null){
+                    // Return the pill immediately if there is one
+                    item = poisonPill;
```

Comment:
I think it might be okay to just return the poison pill even if the pool is already locked - in which case this could move to being the first `if` block.  @jpinsonault, @KevinCorcoran - anything you can think of where this might pose a problem?